Tests/Unit/NetworkServer/ConcentratorDeduplicationTest.cs

```diff
@@ -100,7 +112,32 @@ public void CreateKeyMethod_Should_Produce_Expected_Key_For_UpstreamDataFrames()
             var expectedKey = "EF-BC-21-12-E9-B8-AF-F4-1B-F5-7B-7D-3C-D3-69-5A-84-E3-D8-6C-FC-1A-8E-BC-80-39-10-64-0D-6B-B3-7C";
 
             // act/assert
-            Assert.Equal(expectedKey, ConcentratorDeduplication.CreateCacheKey(this.dataRequest));
+            Assert.Equal(expectedKey, this.concentratorDeduplication.CreateCacheKey(this.dataPayload));
+        }
+
+        [Theory, MemberData(nameof(TestRawMessages))]
+        public void CreateKeyMethod_Should_Include_All_Properties_For_Different_RawMessages(byte[]? rawMessage)
+        {
+            // arrange
+            var mock = new Mock<ConcentratorDeduplication>(MockBehavior.Default, this.cache, NullLogger<ConcentratorDeduplication>.Instance);
+            Memory<byte> actualBuffer = null;
+            _ = mock.Setup(x => x.HashKey(It.IsAny<byte[]>())).Callback<byte[]>(b => actualBuffer = b);
+            this.dataPayload.RawMessage = rawMessage;
+
+            // act
+            _ = mock.Object.CreateCacheKey(this.dataPayload);
+
+            // assert
+            mock.Verify(x => x.HashKey(It.IsAny<byte[]>()), Times.Once);
+            Assert.True(MemoryExtensions.SequenceEqual(this.dataPayload.DevAddr.Span, actualBuffer.Span[0..4]));
+            Assert.True(MemoryExtensions.SequenceEqual(this.dataPayload.Mic.Span, actualBuffer.Span[4..8]));
+            var index = 8;
+            if (this.dataPayload.RawMessage?.Length > 0)
+            {
+                index += this.dataPayload.RawMessage.Length;
+                Assert.True(MemoryExtensions.SequenceEqual(this.dataPayload.RawMessage.AsSpan(), actualBuffer.Span[8..index]));
+            }
+            Assert.True(MemoryExtensions.SequenceEqual(this.dataPayload.Fcnt.Span, actualBuffer.Span[index..])); // implicitly asserts that the length is correct as well
```

Comment:
Why not invoke extensions methods more in naturally as extensions?

```suggestion
            Assert.True(this.dataPayload.DevAddr.Span.SequenceEqual(actualBuffer.Span[0..4]));
            Assert.True(this.dataPayload.Mic.Span.SequenceEqual(actualBuffer.Span[4..8]));
            var index = 8;
            if (this.dataPayload.RawMessage?.Length > 0)
            {
                index += this.dataPayload.RawMessage.Length;
                Assert.True(this.dataPayload.RawMessage.AsSpan().SequenceEqual(actualBuffer.Span[8..index]));
            }
            Assert.True(this.dataPayload.Fcnt.Span.SequenceEqual(actualBuffer.Span[index..])); // implicitly asserts that the length is correct as well
```

Reply:
Thanks (should have realised these are real extension methods not just by name :D) Resolving though because this code was removed.